### DISCUSSION TITLE
Errors starting up with Spring Boot 2.6.6

### DISCUSSION THREAD

**Discussion by therealtry** [evrete/evrete#24] [Q&A] [answered]
Hey people,

I am trying to integrate Evrete into a spring boot 2 microservice and get the following error:

`package org.evrete.spi.minimal.compiled;

public final class Condition1 extends org.evrete.spi.minimal.BaseConditionClass {
    public static final java.lang.invoke.MethodHandle HANDLE;

    static {
        try {
            HANDLE = java.lang.invoke.MethodHandles.lookup().findStatic(Condition1.class, "__$test", java.lang.invoke.MethodType.methodType(boolean.class, org.evrete.api.IntToValue.class));
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static boolean __$testInner(com.fasterxml.jackson.databind.JsonNode var0) {
        return (var0.get("companyCode") != null) ? var0.get("companyCode").asText().equals("MY40") : false;
    }

    public static boolean __$test(org.evrete.api.IntToValue values) {
        return __$testInner((com.fasterxml.jackson.databind.JsonNode) values.apply(0));
    }

    //IMPORTANT LINE BELOW, IT IS USED IN SOURCE/SIGNATURE COMPARISON
    //fields in use: com.logicea.microservices.billing.domain.Invoice/erpMetadata
}

2022-09-08 11:16:51.026 ERROR 23911 --- [           main] org.evrete.runtime.AbstractRuntime       : Compilation error

org.evrete.util.compiler.CompilationException: Compilation error(s): /folder4/Condition1.java:4: error: cannot access org.evrete.spi.minimal.BaseConditionClass
public final class Condition1 extends org.evrete.spi.minimal.BaseConditionClass {
                                                            ^
  bad class file: file:/home/try/workspace/projects/microservices-infrastructure-modular/billing-service/target/billing-service-1.2.4-HOTFIX-UAT.jar!/org/evrete/spi/minimal/BaseConditionClass.class
    unable to access file: java.io.FileNotFoundException: JAR entry org/evrete/spi/minimal/BaseConditionClass.class not found in /home/try/workspace/projects/microservices-infrastructure-modular/billing-service/target/billing-service-1.2.4-HOTFIX-UAT.jar
`

the code looks like this (kotlin):

` init {
        invoiceReferenceSequenceRulesConfiguration.rules?.entries
            ?.forEach { it ->
                val ruleBuilder = knowledge.newRule(it.key)
                val knowledgeLhsBuilder = ruleBuilder.forEach(
                    "\$i", Invoice::class.java
                )

                it.value.forEach { configRule ->
                    knowledgeLhsBuilder.where(configRule.getPredicate()) //this creates an expression like //$i.erpMetadata.get("companyCode").asText().equals("MY40")
                }

                knowledgeLhsBuilder.execute { ctx ->
                    log.info("knowledgeLhsBuilder.execute(): Executing rule for sequence {}", it.key)
                    val invoice: Invoice = ctx.get("\$i")
                    log.info("knowledgeLhsBuilder.execute(): Processing invoice {}", invoice.id)
                    val invoiceRefSequence: String = it.key
                    log.debug("knowledgeLhsBuilder.execute(): invoiceRefSequence: {}", invoiceRefSequence)
                    var sequence = ""

                    retryTemplate.execute<Unit,Exception> {
                        sequence =
                            invoiceReferenceSequenceService.updateCounterForSequence(invoiceRefSequence).currentIndex.toString()

                    }

                    log.debug("knowledgeLhsBuilder.execute(): updated and retrieved sequence {}", it)

                    if (invoiceReferenceSequenceRulesConfiguration.sequenceNameAsPrefix) {
                        sequence = invoiceRefSequence + sequence
                    }
                    invoice.invoiceReference = sequence
                    invoice.issuedDate = Instant.now()

                    log.info(
                        "knowledgeLhsBuilder.execute(): invoice reference for invoice with id {} is now {}",
                        invoice.id,
                        sequence
                    )
                    ctx.delete(invoice)
                }

                log.info("init: Knowledge has been initialized {}", knowledge)

            }
    }`

**Comment by therealtry** [comment]
I understand that the compiled class isn't loaded on the class loader. I tried unzipping the dependency on the `org` folder of the zip and then afterwards it couldn't locate the JsonNode class that is required for some of our rules

**Comment by andbi** [comment]
Hi, it looks like a packaging issue. The "bad class file"  message says that the library classes are missing in the `billing-service-1.2.4-HOTFIX-UAT.jar` What is this jar?  If that's the Spring Boot runnable jar, then Evrete libs should be located in the `/BOOT-INF/lib` folder of the jar, along with other dependencies.

**Reply by therealtry** [reply]
Hi @andbi thanks for the quick response. It indeed is in there:

![image](https://user-images.githubusercontent.com/18594338/189087599-5ae1974d-c78c-4cdb-a4b3-df1e6fc9213a.png)

**Reply by therealtry** [reply]
It really is a packaging issue because it runs flawlessly from the IDE (IntelliJ). Btw, I think it's a great library from what I've seen so far. Kudos!

**Reply by andbi** [reply]
I'm using it in with Spring boot too, but I never tried to initialize it on the configuration level where classloading might be different, especially when SPIs are involved. As a workaround, you can try setting it up in a service and see if the problem is gone.

**Reply by therealtry** [reply]
Ok will try that and let you know! Thanks!

**Reply by therealtry** [reply]
I tried setting it in a service PostConstruct. I debugged to a point where I reached this class:

`class JcCompiler extends SourceCompiler {
    private final ProtectionDomain protectionDomain;

    JcCompiler(ProtectionDomain protectionDomain) {
        this.protectionDomain = protectionDomain;
    }

    Class<?> compile(ClassLoader classLoader, String source) throws CompilationException {
        ServiceClassLoader cl;
        if(classLoader instanceof ServiceClassLoader) {
            cl = (ServiceClassLoader) classLoader;
        } else {
            cl = new ServiceClassLoader(classLoader, protectionDomain);
        }
        return compile(source, cl);
    }

}`

after the invocation of the ServiceClassLoader constructor after the else, there are no classes loaded anymore. Up until that point, I could find the classes required with evaulate expression

**Reply by therealtry** [reply]
My last comment was invalid since I was using the protected `findClass` function instead of the `loadClass`. So far my attempts have failed. Would you be able to provide some code snippets where you showcase evrete configured with spring boot? Thank you for your responses and time so far. I really appreciate them

**Reply by andbi** [reply]
Sure, here is a fragment of my abstract class which is extended by various `@Service` components:
```java
public abstract class AbstractComponent {
    private static final KnowledgeService KNOWLEDGE_SERVICE = new KnowledgeService();

    @Transactional
    public void process(String arg) {
        Knowledge kn = KNOWLEDGE_SERVICE.newKnowledge("JAVA-CLASS", AnalyzeRulesetEn.class);
        StatelessSession session = kn.newStatelessSession();
        // Session config & inserts
        session.fire();
    }
}
```

My Maven dependencies are simple too:
```xml
<dependency>
    <groupId>org.evrete</groupId>
    <artifactId>evrete-core</artifactId>
    <version>2.2.01</version>
</dependency>

<dependency>
    <groupId>org.evrete</groupId>
    <artifactId>evrete-dsl-java</artifactId>
    <version>2.2.01</version>
</dependency>
```

**Reply by andbi** [reply]
Wait, my app runs as war on standalone Tomcat instances, let me check if works the same way when deployed as runnable jar.

**Reply by therealtry** [reply]
My config looks pretty much the same, I was about to say

**Reply by andbi** [reply]
I got the same error when running as boot jar:-( On it.

**Reply by therealtry** [reply]
you are a wizard

**Reply by andbi** [reply]
https://github.com/evrete/evrete/issues/25

**Reply by andbi** [reply]
Bug fixed in version 2.2.02, it should be already available in Maven Central.

**Reply by therealtry** [reply]
You are a lifesaver!

**Comment by andbi** [comment, marked as answer]
Many thanks for bringing up this issue. I'm just wondering how many developers must've faced it too and abandoned the library because of it.

**Reply by therealtry** [reply]
Who knows? But this is going to production so if I have any feedback I'll shoot you an email! Again, thanks a million!
